aspnetcore/release-notes/aspnetcore-3.0.md

```diff
@@ -134,7 +134,229 @@ For more information, see <xref:blazor/index>.
+        {
+            return currentUsername.Equals("bob42@jabbr.net", StringComparison.OrdinalIgnoreCase);
+        }
+
+        return currentUsername.EndsWith("@jabbr.net", StringComparison.OrdinalIgnoreCase));
+    }
+}
+```
+
+In the preceding code, `DomainRestrictedRequirement` serves as a custom `IAuthorizationRequirement`. Because the `HubInvocationContext` resource parameter is being passed in, the internal logic can:
+
+* Inspect the context in which the Hub is being called.
+* Make decisions on allowing the user to execute individual Hub methods.
+
+Individual Hub methods can be decorated with the name of the policy the code checks at run-time. As clients attempt to call individual Hub methods, the `DomainRestrictedRequirement` handler runs and controls access to the methods. Based on the way the `DomainRestrictedRequirement` controls access:
+
+* All logged-in users can call the `SendMessage` method.
+* Only users who have logged in with a `@jabbr.net` email address can view users’ histories.
+* Only `bob42@jabbr.net` can ban users from the chat room.
+
+```csharp
+[Authorize]
+public class ChatHub : Hub
+{
+    public void SendMessage(string message)
+    {
+    }
+
+    [Authorize("DomainRestricted")]
+    public void BanUser(string username)
+    {
+    }
+
+    [Authorize("DomainRestricted")]
+    public void ViewUserHistory(string username)
+    {
+    }
+}
+```
+
+Creating the `DomainRestricted` policy might involve:
+
+* In *Startup.cs*, adding the new policy.
+* Provide the custom `DomainRestrictedRequirement` requirement as a parameter.
+* Registering `DomainRestricted` with the authorization middleware.
+
+```csharp
+services
+    .AddAuthorization(options =>
+    {
+        options.AddPolicy("DomainRestricted", policy =>
+        {
+            policy.Requirements.Add(new DomainRestrictedRequirement());
+        });
+    });
+```
+
+SignalR hubs use [Endpoint Routing](xref:fundamentals/routing). SignalR hub connection was previously done explicitly:
+
+```csharp
+app.UseSignalR(routes =>
+{
+    routes.MapHub<ChatHub>("hubs/chat");
+});
+```
+
+In the previous version, developers needed to wire up controllers, Razor pages, and hubs in a variety of different places. Explicit connection results in a series of nearly-identical routing segments:
+
+```csharp
+app.UseSignalR(routes =>
+{
+    routes.MapHub<ChatHub>("hubs/chat");
+});
+
+app.UseRouting(routes =>
+{
+    routes.MapRazorPages();
+});
+```
+
+SignalR 3.0 hubs can be routed via endpoint routing. With endpoint routing, typically all routing can be configured in `Startup.Configure`:
+
+```csharp
+app.UseRouting(routes =>
+{
+    routes.MapRazorPages();
+    routes.MapHub<ChatHub>("hubs/chat");
+});
+```
+
+ASP.NET Core 3.0 SignalR added:
+
+* Streaming support, which enables streaming return values from server-side methods. This is useful for when fragments of data will come in over a period of time.
```

Comment:
This is not new in 3.0...

Reply:
> This is not new in 3.0...

@BrennanConroy  I'll remove that from #14240. This has been merged into #14240 so leave comments in #14240.  I'll remove the following from `ASP.NET Core 3.0 SignalR added:` to
`## Generic Host`

That is, the remaining SignalR content.

Reply:
ack. i globbed server-to-client in there by accident. @Rick-Anderson - you got this change or need me to hop on it?